docs/Toy/toy/compiler.py

```diff
@@ -124,20 +124,32 @@ def transform(
     DeadCodeElimination().apply(ctx, module_op)
     ReconcileUnrealizedCastsPass().apply(ctx, module_op)
 
-    DeadCodeElimination().apply(ctx, module_op)
-
     module_op.verify()
 
     if target == "riscv":
         return
 
+    CanonicalizePass().apply(ctx, module_op)
+
+    module_op.verify()
+
+    if target == "riscv-opt":
+        return
+
     RISCVRegisterAllocation().apply(ctx, module_op)
 
     module_op.verify()
 
     if target == "riscv-regalloc":
         return
 
+    CanonicalizePass().apply(ctx, module_op)
```

Comment:
optimises things that need register allocation, like redundant moves etc

Reply:
```suggestion
    # Perform optimizations that depend on register allocation (e.g.: redundant moves):
    CanonicalizePass().apply(ctx, module_op)
```